### ISSUE TITLE
Support for Tensorboard logger

### ISSUE THREAD

**Issue by aleSuglia** [facebookresearch/nle#92]
## 🚀 Feature

Use Tensorboard for logging statistics of the agent during training instead of plain TSV files.

## Motivation

Tensorboard is the common way of plotting data used by most of the Deep Learning framework based on PyTorch (some examples are PyTorch-lightning, AllenNLP, etc.). In addition, it is currently included in the official PyTorch library: https://pytorch.org/docs/stable/tensorboard.html

## Pitch

Don't get me wrong, I love the geeky terminal plotting feature. However, I believe it would be better to use something more versatile like the Tensorboard. I would be happy to contribute to this. I basically have a working implementation already that requires a few steps of debugging.

Let me know in case you would be interested in supporting it.

**Comment by edran** [comment]
Hi @aleSuglia -- thank you for opening the issue and for the suggestion.

My personal preference is that we keep the agent as simple as possible, since it's really only meant to be reference implementation showcasing how to do basic things with NLE.

Furthermore, particularly on the issue of tensorboard, I appreciate that it is popular, but (a) it pulls a lot of unnecessary packages with it, and (b) since none of us in the dev team really use it, we would struggle to maintain it.

As an asied, the beauty of using a simple logging system based on tsv / csv files is that a user can quickly fork the code and swap them out for their choice of logging platform.

@heiner @rockt what do you think?

**Comment by aleSuglia** [comment]
I see your point. I believe that it would be very easy to make this configurable by creating an abstract `Logger` class that can be instantiated by the user depending on a CLI option. By default, you will always get a `CSVLogger` while `TensorboardLogger` can be optionally enabled.

**Comment by heiner** [comment]
Hey Alessandro,

First of all, thanks for your interest in NLE and your suggestion!

I have to say I agree with @edran on this one.

(I also have an idiosyncratic dislike of TensorBoard in particular, but that's beside the point :))

Correct me if I'm wrong, but this feature requests is about `agent.py` in particular? If so, my feelings are: (1) This agent implementation is an example (e.g., it does not and will never reproduce our paper results) and (2) if you are basing your work on this agent implementation, you would presumably take `agent.py` and modify it anyway -- e.g., substitute your own neural network model. While doing so, it should be possible to implement your favourite logging mechanism?

**Comment by aleSuglia** [comment]
@heiner Yes, my comment was mainly about the agent implementation. I thought it was meant to be the baseline/official implementation based on the NeurIPS paper that's why I was suggesting to adopt something more wide-spread. I thought it would have been the starting point for a lot of people playing with NLE. I feel like we can then close this issue then. Thanks to both for taking the time to answer :)

P.S. At this stage, my question becomes: how do we reproduce your results? If you prefer, I can send an e-mail so that you can provide me with more details about the implementation? I would be happy to implement them myself following your guidance.

Thanks,
Alessandro

**Comment by heiner** [comment]
Hey Alessandro,

That's a very sensible question and we have been a bit slow about this.

We're working on preparing a code-dump of our (currently very messy) codebase which we used for the NLE paper. Since NLE itself was developed and changed while we produced these results there are a few incongruities that we'll have to resolve, and as an engineer I'd rather release something somewhat I can stand behind.

You are also not the first to ask, Victor Zhong and others have asked at https://github.com/facebookresearch/nle/issues/90 so we are very aware of this request and we will release code that reproduces the paper (or perhaps gets slightly better results) in the future.

As an aside by the way: In research code I'm responsible for I'd fight tooth and nail against a suggestion like "creating an abstract `Logger` class". The reason is that research code is complex and messy enough as it is and should not in my opinion have any more "concepts" for which the reader needs to understand the author's way of thinking than absolutely necessary. Hence, no "Loggers". In my experience, fork&hack something concrete is a much better way.
